### PR TITLE
Booking logic

### LINKED ISSUE #21: Implement Booking Logic
Checks to implement:

- [x] User in Org
- [x] Round to minutes
- [x] Start time < end time
- [x] No conflicting booking
- [x] Duration < x hours or isAdmin
- [x] No stacking booking

### PR DESCRIPTION
**Pull request by SeekSaveServe**
Adds check functions and test cases for booking logic

Each check function corresponds to a type of error that could occur when a booking is submitted

1. checkUserInOrg: ERROR - User cannot book for an organisation he is not in
2. checkStartEndTime: ERROR - Please ensure that you have selected the correct duration and that bookings are in multiples of the slot duration
3. checkBookingPrivilege: ERROR - Your booking duration exceeds your booking privilege. Please look for an admin for long bookings
4. checkConflictingBooking: Error - Your booking overlaps with another booking in the same venue
5. checkStackedBooking: Error - Your organisation has stacked too many bookings!

I think check 2 can be shifted to the front end. We can also consider customizing the error message with a redirect to a `booking rules` page on the front end

closes #21

### CODE REVIEW

**Comment by misterpuffin** [comment]
btw i got the same npm error initially as i was using v16! had to downgrade to v14 and rebuild the package-lock with `npm i --package-lock-only`

File: src/middlewares/checks.ts [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+ * @param start
+ * @param end
+ * @returns true if  there is no violation
+ */
+export function checkStartEndTime(start: Date, end: Date): boolean {
+  return (
+    (convertDateToMinutes(end) - convertDateToMinutes(start)) /
+      DURATION_PER_SLOT >=
+      MIN_SLOTS_PER_BOOKING &&
+    (convertDateToMinutes(end) - convertDateToMinutes(start)) %
```
Comment (misterpuffin) [inline src/middlewares/checks.ts:72]: Do we want to check that the start/end times are in intervals of for eg. 30 mins? or would the input be sanitised already. 

So for eg. if convertDateToMinutes(end) % DURATION_PER_SLOT != 0 we return false?

File: src/middlewares/checks.ts [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+/**
+ * Checks if the duration of the booking is within the booking privilege of the organisation
+ *
+ * @param booking
+ * @returns true if duration exceeds privilege
+ */
+export async function checkBookingPrivelege(booking: BookingPayload) {
+  const { start, end, userId } = booking
+  const userOnAdmin = await prisma.userOnOrg.findMany({
```
Comment (misterpuffin) [inline src/middlewares/checks.ts:86]: perhaps we can abstract this out to an checkIsAdmin function since this seems like it would be useful

**Review by misterpuffin — APPROVED** [review]
overall lgtm thanks for doing the tests as well :D

File: src/middlewares/checks.ts (hunk 1 above)
Comment (SeekSaveServe) [reply src/middlewares/checks.ts:72]: ah i see, i only checked that the duration is in multiples of 30 mins. Good catch!

File: src/middlewares/checks.ts (hunk 2 above)
Comment (SeekSaveServe) [reply src/middlewares/checks.ts:86]: sure!

**Pull request merged**